stored_requests/events/postgres/database.go

```diff
@@ -0,0 +1,177 @@
+	if err != nil {
+		glog.Warningf("Failed to fetch all Stored %s data from the DB: %v", e.cfg.RequestType, err)
+		// e.saves <- events.Save{} //TODO: do we need to do this?
+		return err
+	}
+	defer func() {
+		if err := rows.Close(); err != nil {
+			glog.Warningf("Failed to close the Stored %s DB connection: %v", e.cfg.RequestType, err)
+		}
+	}()
+	if err := e.sendEvents(rows); err != nil {
+		glog.Warningf("Failed to load all Stored %s data from the DB: %v", e.cfg.RequestType, err)
+		// e.saves <- events.Save{} //TODO: do we need to do this?
+		return err
+	} else {
+		e.lastUpdate = thisTimeInUTC
+	}
+	return nil
+}
+
+func (e *PostgresEventProducer) fetchDelta() error {
+	thisTimeInUTC := e.time.Now().UTC()
+
+	timeout := e.cfg.CacheUpdateTimeout * time.Millisecond
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	rows, err := e.cfg.Db.QueryContext(ctx, e.cfg.CacheUpdateQuery, e.lastUpdate)
+
+	if err != nil {
+		glog.Warningf("Failed to fetch updated Stored %s data from the DB: %v", e.cfg.RequestType, err)
+		return err
+	}
+	defer func() {
+		if err := rows.Close(); err != nil {
+			glog.Warningf("Failed to close the Stored %s DB connection: %v", e.cfg.RequestType, err)
+		}
+	}()
+	if err := e.sendEvents(rows); err != nil {
+		glog.Warningf("Failed to load updated Stored %s data from the DB: %v", e.cfg.RequestType, err)
+		return err
+	} else {
+		e.lastUpdate = thisTimeInUTC
+	}
+	return nil
+}
+
+// sendEvents reads the rows and sends notifications into the channel for any updates.
+// If it returns an error, then callers can be certain that no events were sent to the channels.
+func (e *PostgresEventProducer) sendEvents(rows *sql.Rows) (err error) {
+	storedRequestData := make(map[string]json.RawMessage)
+	storedImpData := make(map[string]json.RawMessage)
+
+	var requestInvalidations []string
+	var impInvalidations []string
+
+	for rows.Next() {
+		var id string
+		var data []byte
+		var dataType string
+
+		// Beware #338... we really don't want to save corrupt data
+		if err := rows.Scan(&id, &data, &dataType); err != nil {
+			return err
+		}
+
+		switch dataType {
+		case "request":
+			if len(data) == 0 || bytes.Equal(data, []byte("null")) {
+				requestInvalidations = append(requestInvalidations, id)
+			} else {
+				storedRequestData[id] = data
+			}
+		case "imp":
+			if len(data) == 0 || bytes.Equal(data, []byte("null")) {
+				impInvalidations = append(impInvalidations, id)
+			} else {
+				storedImpData[id] = data
+			}
+		default:
+			glog.Warningf("Stored Data with id=%s has invalid type: %s. This will be ignored.", id, dataType)
+		}
+	}
+
+	// Beware #338... we really don't want to save corrupt data
+	if rows.Err() != nil {
+		return rows.Err()
+	}
+
+	if len(storedRequestData) > 0 || len(storedImpData) > 0 {
+		e.saves <- events.Save{
+			Requests: storedRequestData,
+			Imps:     storedImpData,
+		}
+	}
+
+	//TODO: do we really need to check that this isn't a fetch all here?
```

Comment:
Can you please add more context about the question in this comment?

Reply:
Removed this comment with no changes to the code. I think that we should keep the check in here to ensure that cache invalidations only occur during a fetch delta. When we fetch all, the cache should be empty so there's no reason to perform any cache invalidations, even if the data returned is empty or `null`.